### PR TITLE
fix: Jump links styles and browser fixes

### PR DESCRIPTION
**Pull request by castastrophe**
### Related issues

- #1521
- #1343 
- #1258 
- #1162 
- #1028

### What has changed and why

- Add support to pfe-accordion-header for slotted content to be read in and rendered (this allows us to use pfe-accordion inside another template and pass a slot into the pfe-accordion-header)
- Add PfeAccordion to pfe-jump-links-nav as a dependency
- Add support for the following events:
  - `pfe-jump-links-panel:active-navItem`: when a new item in the nav becomes active, this event is fired for analytics; this is no longer handled by the panel but the event name persists for backwards compatibility
  - `pfe-jump-links-panel:change`: fired when the panel or section elements change; used by the autobuilder; calls _panelChangedHandler
  - `pfe-jump-links-nav:stuck`
  - `resize`: on page resize, the rebuild method is called
  - `scroll`: handles updating the active states when the page is scrolled
  - `keyup`: this is a stub and needs to be addressed at a later date (@TODO open issue for this)
- Added the following attributes to `pfe-jump-links-nav`:
  - `offset`: the manually set distance from the top of the viewport to the top of the navigation element; this is an override and does not account for pfe-navigation when used.
  - `mobile-breakpoint`: max-width for when to switch to the pfe-accordion for mobile rendering
  - `accordion-collapse-timing`: how long to wait after clicking a child element before collapsing the open accordion
  - `stuck`: represents the stickiness state
  - `no-header`: boolean attribute to allow you to opt-out of rendering the header section (header becomes screen-reader only)
- Added the following attributes to `pfe-jump-links-panel`:
  - `spacers`: opt-in attribute for injecting the spacer elements; not recommended but allows backward compatibility

Going forward, the only required component to add to the page is `pfe-jump-links-nav`.  Panel is an optional wrapper to make targeting more efficient.  All functionality is being handled inside the nav element.  Panel and sections can be manually defined via JavaScript API call to the `pfe-jump-links-nav` component.

#### Test cases

- [x] *Initialization*
  - [URL](https://deploy-preview-1652--patternfly-elements.netlify.app/elements/pfe-jump-links/demo/)
  - Validate that on page load, the standard heading is selected in the top horizontal navigation and the Section 1 region is selected in the first visible vertical jump links.
- [x] *Stickiness*
  - [URL](https://deploy-preview-1652--patternfly-elements.netlify.app/elements/pfe-jump-links/demo/)
  - On scroll, the horizontal and vertical elements should stick to the top of the viewport while it's content is visible; it should account for the height of pfe-navigation if it exists and any horizontal sticky elements
  - Edge case: If 2 horizontal sticky elements are present, the one later in the DOM will overlap the other one.
- [x] *Active state*
  - [URL](https://deploy-preview-1652--patternfly-elements.netlify.app/elements/pfe-jump-links/demo/)
  - On scroll, the active state in the navigation should update when the section scrolls close to the top of the viewport.
- [x] *Click event*
  - [URL](https://deploy-preview-1652--patternfly-elements.netlify.app/elements/pfe-jump-links/demo/)
  - On clicking a navigation item on either horizontal or vertical nav, the page should smooth scroll to that section so that the top of the section is visible close to the top of the viewport (accounting for sticky elements) with at least a 20px breathing space (unless offset is manually set in the mark-up)
  - No in-between sections should show as active while the scroll is happening
  - Scroll controlled active states should resume after the scrolling event is complete

#### Browser requirements

Your component should work in all of the following environments:

- [x] Latest 2 versions of Edge
- [x] Internet Explorer 11 (should be useable, not pixel perfect)
- [x] Latest 2 versions of Firefox (one on Mac OS, one of Windows OS)
- [x] Firefox 78 (or latest version for Red Hat Enterprise Linux distribution)
- [x] Latest 2 versions of Chrome (one on Mac OS, one of Windows OS)
- [x] Latest 2 versions of Safari
- [x] Android mobile device (such as the Galaxy S9)
- [x] Apple mobile device (such as the iPhone X)
- [x] Apple tablet device (such as the iPhone Pro)

### Ready-for-merge Checklist

- [x] Expected files: all files in this pull request are related to one request or issue (no stragglers or scope-creep).
- ~~[ ] Tests have been updated to cover these changes. **TODO** return to this after code freeze~~
- [x] Browser testing passed.
- [x] Changelog updated.
- [x] Documentation (README.md, WHY.md, etc.) updated or added.
- [ ] Link to the demo recording: []()
- [ ] Approved by designer.

### Merging

Please **squash** when merging and ensure your commit message uses [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) formatting.

**Be sure to share your updates with the [patternfly-elements-contribute@redhat.com](mailto:patternfly-elements-contribute@redhat.com) mailing list!**

### CODE REVIEW

File: elements/pfelement/src/pfelement.js [hunk 1]
```diff
@@ -247,23 +247,22 @@ class PFElement extends HTMLElement {
       return;
     }
 
-    switch (typeof name) {
-      case "string":
-        return (
-          [...this.children].filter((child) => child.hasAttribute("slot") && child.getAttribute("slot") === name)
+    if (typeof name === "string") {
```
Comment (castastrophe) [inline elements/pfelement/src/pfelement.js:250]: @mwcz Funnily enough, I picked up this jump links branch and it needed the fix here that we talked about this morning!

**Comment by zhawkins** [comment]
branch testing looks good 👍

File: elements/pfe-tabs/src/polyfills--pfe-tabs.js [hunk 2]
```diff
@@ -1,51 +1,3 @@
-// @POLYFILL  Array.prototype.find
```
Comment (heyMP) [inline elements/pfe-tabs/src/polyfills--pfe-tabs.js:1]: pfe-tabs.js is using find() in two locations.  Do you think we still need the polyfill?

- https://github.com/patternfly/patternfly-elements/blob/7c25ac067d6bd71ed019a9c7fd67945d2c9208ea/elements/pfe-tabs/src/pfe-tabs.js#L398
- https://github.com/patternfly/patternfly-elements/blob/7c25ac067d6bd71ed019a9c7fd67945d2c9208ea/elements/pfe-tabs/src/pfe-tabs.js#L445

File: elements/pfelement/src/pfelement.js [hunk 3]
```diff
@@ -1075,8 +1090,27 @@ class PFElement extends HTMLElement {
   static get cascadingProperties() {
     return this._getCache("cascadingProperties");
   }
+
+  /**
+   * Breakpoint object mapping human-readable size names to viewport sizes
+   * To overwrite this at the component-level, include `static get breakpoint` in your component's class definition
+   * @returns {Object} keys are t-shirt sizes and values map to screen-sizes (sourced from PF4)
+   */
+  static get breakpoint() {
```
Comment (heyMP) [inline elements/pfelement/src/pfelement.js:1099]: We have breakpoints! 🔥

File: elements/pfe-tabs/src/polyfills--pfe-tabs.js (hunk 2 above)
Comment (heyMP) [reply elements/pfe-tabs/src/polyfills--pfe-tabs.js:1]: nvm.  I see that you added it to Pfelement. 👍

File: scripts/gulpfile.factory.js [hunk 4]
```diff
@@ -134,29 +134,38 @@ module.exports = function factory({
...
   const fetchHtml = url => {
+    if (!url) return null;
+
     // Check for the html template
-    if (url && fs.existsSync(path.join(paths.source, url))) {
+    if (fs.existsSync(path.join(paths.source, url))) {
       // Returns a string with the cleaned up HTML
-      return decomment(
-        fs
+      // decomment(
```
Comment (heyMP) [inline scripts/gulpfile.factory.js:153]: Was this to get something to work or is it more for better dev experience?

File: elements/pfe-jump-links/src/pfe-jump-links-nav.js [hunk 5]
```diff
@@ -30,16 +66,46 @@ class PfeJumpLinksNav extends PFElement {
...
+      mobileBreakpoint: {
+        title: "Mobile breakpoint (max-width)",
+        type: String,
+      },
+      accordionCollapseTiming: {
+        title: "Number of ms to wait before collapsing the accordion on click",
+        type: Number,
+        default: 750,
+      },
+      // Reflects if the nav is stuck in place
```
[the remark this answers is not in the document]
Comment (castastrophe) [reply elements/pfe-jump-links/src/pfe-jump-links-nav.js:101]:
```suggestion
      // Reflects if the nav is stuck in place
      // @TODO note this in the documentation
```

File: scripts/gulpfile.factory.js (hunk 4 above)
Comment (castastrophe) [reply scripts/gulpfile.factory.js:153]: Yikes, I don't even remember.  let me add it back and see what breaks

Comment (castastrophe) [reply scripts/gulpfile.factory.js:153]: Oh I remember now!  It's stripping out comments and that breaks the commented out code in the pfe-jump-links.js file

Comment (heyMP) [reply scripts/gulpfile.factory.js:153]: 👍

File: elements/pfe-jump-links/src/pfe-jump-links-nav.js [hunk 6]
```diff
@@ -917,20 +917,20 @@ class PfeJumpLinksNav extends PFElement {
...
-        // this prevents the height from taking into account the open accordion tray
-        if (this.isMobile) {
-          const accordionHeader = this.shadowRoot.querySelector("pfe-accordion-header");
-          height = accordionHeader.getBoundingClientRect().height;
-        }
+      // Account for it's height as well
+      // this.offsetVar does not account for this because this only affects scrolling to sections
+      let height = 0;
+
+      if (this.horizontal) height = this.getBoundingClientRect().height;
```
Comment (heyMP) [inline elements/pfe-jump-links/src/pfe-jump-links-nav.js:925]: Just verifying that we don't need to check that we are in a "stuck" state here. ie:

```
if (this.stuck) {
  if (this.horizontal) height = this.getBoundingClientRect().height;
}
```

Comment (castastrophe) [reply elements/pfe-jump-links/src/pfe-jump-links-nav.js:925]: I don't think so because after scrolling, the mobile nav would be sticky anyway right?

Comment (heyMP) [reply elements/pfe-jump-links/src/pfe-jump-links-nav.js:925]: Makes sense 👍

**Review by heyMP — APPROVED** [review]
🎉

**Pull request merged**